### PR TITLE
Normalize Chutes qwen <think> CoT into reasoning_content (#779)

### LINKED ISSUE #779: Chutes qwen models leak raw <think> CoT into content; reasoning_content null, reasoning_tokens=0
Found in the Chutes prod-readiness bug-hunt (staging).

## Symptom
On the **Chutes** path, `qwen/qwen3-32b` (and almost certainly `qwen/qwen3-235b-a22b-thinking-2507`) returns the model's chain-of-thought as literal `<think>…</think>` inside `message.content`, with `reasoning_content` absent/null and `usage.reasoning_tokens=0` despite hundreds of reasoning tokens.

**Inconsistent across models on the same API:** `minimax/minimax-m2.5` (Chutes) correctly populates `reasoning_content`, and NEAR-primary `zai-org/GLM-5.1-FP8` correctly separates `reasoning_content` + reports `reasoning_tokens`. So the think-tag reasoning parser is applied for NEAR / some models but **not** for the qwen Chutes models.

## Impact (3 ways)
1. **Quality/UX + CoT privacy** — clients rendering `content` verbatim show raw chain-of-thought to end users.
2. **Broken `reasoning_content`** — clients reading the dedicated field get nothing.
3. **Billing/metering drift** — `reasoning_tokens=0` while `completion_tokens` includes the CoT → mismetered.

## Repro
`POST /v1/chat/completions {"model":"qwen/qwen3-32b","messages":[{"role":"user","content":"What is 2+2? Reply with just the number."}],"max_tokens":2000,"temperature":0}` → `content` starts `<think>\nOkay, so the user…`, ends `</think>\n\n4`; `reasoning_content` absent; `reasoning_tokens=0`, `completion_tokens=285`.

## Fix direction
Apply the think-tag reasoning parser on the Chutes response path for the qwen models (route `<think>…</think>` to `reasoning_content`, count `reasoning_tokens`) — mirror what minimax/GLM-5.1 already get. Likely in the Chutes provider's response handling (it already re-serializes on the canonical-id rewrite path) or the shared reasoning-extraction layer.

### PR DESCRIPTION
**Pull request by Evrard-Nil**
## Summary
Closes #779.

On the **Chutes** path, the qwen models (`qwen/qwen3-32b`, `qwen/qwen3-235b-a22b-thinking-2507`) return chain-of-thought as a literal `<think>…</think>` block inside `message.content` with `reasoning_content` null, whereas `minimax` (Chutes) and `zai-org/GLM-5.1-FP8` (NEAR) correctly populate `reasoning_content`. The leak has three effects:
1. **CoT privacy / UX** — clients rendering `content` verbatim show raw chain-of-thought.
2. **Broken `reasoning_content`** — clients reading the dedicated field get nothing.
3. **Metering** — `reasoning_tokens=0` while `completion_tokens` includes the CoT.

## What this does
On the **non-stream** Chutes response path (composed after #782's `transform_response_json`), for each `choices[].message`:
- if `message.content` is a string that **begins** with a well-formed `<think>…</think>` block AND `reasoning_content` is absent/empty → move the inner think text into `message.reasoning_content` and strip the block (plus the whitespace separating it from the answer) from `message.content`.

Robustness (matches the issue's requirements, all covered by unit tests):
- no think block → content left untouched;
- already-populated `reasoning_content` → message left untouched (**minimax is not clobbered**);
- malformed/unclosed `<think>` with no `</think>` → content left untouched (**never lose content**);
- think block not at the very start (answer text first) → left untouched (avoids false positives where a model legitimately discusses `<think>` in its answer);
- non-string content (multimodal array / null) → left untouched.

The normalizer mirrors the existing #780/#782 design (operates on the JSON `serde_json::Value`, touching only the named fields) and reuses the chutes module's test helpers/conventions. `reasoning_content` is the same typed field minimax/GLM-5.1 already surface (`ChatResponseMessage.reasoning_content`).

## `reasoning_tokens` — documented limitation
`usage.completion_tokens_details.reasoning_tokens` is **left untouched**. The upstream body carries no separate reasoning-token count, and re-tokenizing to split `completion_tokens` would require the model tokenizer and risk a **fabricated** number — explicitly out of scope per the issue. Token-accurate reasoning metering for these models is a follow-up (it needs an upstream count or a tokenizer pass).

## Scope: non-stream only
Think tags span SSE chunks, so a correct stream parser must be **stateful across chunks**. Shipping a per-chunk think parser would be half-correct (a `<think>` opened in one chunk and closed in another would be missed or split wrong). The stream path's per-`data:`-chunk `transform_response_json` deliberately does **not** run the normalizer. Chutes streaming is opt-in/experimental (`CHUTES_ENABLE_STREAMING`, gated off by default — no authenticated frame ordering), so the non-stream fix covers the production path. **Follow-up:** stateful stream-side `<think>` reasoning extraction.

## Tests
New unit tests in the chutes module (all green; full `attested::chutes` suite passes):
- `split_leading_think_block_*` — extraction + leading-whitespace + reasoning-only + the not-applicable cases (no tag / mid-string / unclosed / empty);
- `normalize_think_reasoning_*` — block moved to `reasoning_content` + stripped from `content`; existing `reasoning_content` not clobbered; empty reasoning treated as unset; no-think content intact; malformed think intact; non-string content ignored; **usage untouched** (no fabricated reasoning-token count).

`cargo fmt -p inference_providers --check`, `cargo clippy -p inference_providers --all-targets --all-features -- -D warnings`, and `cargo test -p inference_providers --lib attested::chutes` all pass.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by PierreLeGuen — APPROVED** [review]
Clean, well-scoped change: the `<think>` block normalizer is applied only on the Chutes non-stream path after the existing sanitize/canonical-id transform, and the fail-safe behavior is right throughout. No blocking findings.

Verified along the way:

- Mutating the response bytes is safe here — it happens after E2EE decryption and Chutes carries no separate signature (`supports_chat_signatures == false`, `crates/inference_providers/src/attested/chutes/mod.rs:697-698`), so no attestation invariant is broken.
- The typed `ChatCompletionResponse` and `raw_bytes` derive from the same normalized JSON, so they can't diverge; usage tracking reads only `response.usage`, which the normalizer leaves untouched.
- `reasoning_content` is the established field across the codebase (`crates/inference_providers/src/models.rs:34,713`) and is not stripped by `strip_internal_response_fields`, so nothing is lost on re-parse.
- Edge cases behave sensibly: unclosed `<think>` leaves content intact, an already-populated `reasoning_content` is never overwritten (minimax/GLM unaffected), and non-string content is skipped — all covered by the new tests.

Checks run locally: `cargo test -p inference_providers --lib attested::chutes` (79 passed, 1 ignored live probe), `cargo clippy -p inference_providers --all-targets -- -D warnings` (clean), `cargo fmt -p inference_providers --check` (clean); full e2e suite skipped as it needs PostgreSQL and the change is confined to the Chutes provider.

File: crates/inference_providers/src/attested/chutes/mod.rs [hunk 1]
```diff
@@ -494,6 +494,91 @@ fn transform_response_json(bytes: &[u8], canonical: Option<&str>) -> Option<Vec<
...
+        // Only act on a plain-string content (the qwen shape). A content that is an
+        // array of parts, null, or absent is left exactly as-is.
+        let Some(content) = message.get("content").and_then(Value::as_str) else {
+            continue;
+        };
+        let Some((reasoning, answer)) = split_leading_think_block(content) else {
+            continue;
+        };
+        message.insert("content".to_string(), Value::String(answer));
+        message.insert("reasoning_content".to_string(), Value::String(reasoning));
```
[the remark this answers is not in the document]
Comment (Evrard-Nil) [reply crates/inference_providers/src/attested/chutes/mod.rs]: Good catch — fixed in e635f5a5. An empty or whitespace-only `<think></think>` block is now stripped from `content` but no longer sets `reasoning_content` (and clears any pre-existing empty one), consistent with the "empty == unset" contract applied to the provider's own field. Added a test covering empty and whitespace-only blocks.

File: crates/inference_providers/src/attested/chutes/mod.rs [hunk 2]
```diff
@@ -616,6 +701,27 @@ impl InferenceProvider for Provider {
...
+        // or ever losing content on a malformed tag. Runs ONLY on this non-stream path
+        // (think tags span SSE chunks — the stream path is intentionally not normalized;
+        // see `normalize_think_reasoning`). This is applied AFTER `transform_response_json`
+        // so it composes with the canonical-id rewrite + #780 internal-field strip, and
+        // is the byte payload the route uses for hashing/passthrough.
+        let raw_bytes = {
+            let mut v: Value = serde_json::from_slice(&raw_bytes).map_err(|e| {
+                CompletionError::CompletionError(format!("parse sanitized response: {e}"))
+            })?;
+            if let Some(o) = v.as_object_mut() {
```
[the remark this answers is not in the document]
Comment (Evrard-Nil) [reply crates/inference_providers/src/attested/chutes/mod.rs:729]: Declining this refactor: the double parse/serialize is intentional and was reviewed and approved as-is. The normalizer runs only on the non-stream Chutes path (already one of the lighter paths), and folding it into `transform_response_json` would couple a Chutes-qwen-specific concern into the shared sanitize helper that the per-chunk stream path also calls — exactly the path we must NOT normalize. Keeping it as a distinct, clearly-scoped pass is the safer trade-off here over shaving one parse on a non-hot path. Happy to revisit if profiling shows it matters.

File: crates/inference_providers/src/attested/chutes/mod.rs [hunk 3]
```diff
@@ -1200,6 +1306,177 @@ mod tests {
...
+        assert!(
+            obj["usage"].get("reasoning_tokens").is_none()
+                && obj["usage"].get("completion_tokens_details").is_none(),
+            "no reasoning-token count fabricated"
+        );
+        // But the content/reasoning_content normalization still happened.
+        assert_eq!(obj["choices"][0]["message"]["content"], "a");
+        assert_eq!(obj["choices"][0]["message"]["reasoning_content"], "r");
+    }
+
```
[the remark this answers is not in the document]
Comment (Evrard-Nil) [reply crates/inference_providers/src/attested/chutes/mod.rs:1510]: Declining the typed-struct/borrowed-slice refactor for this PR. Operating on `serde_json::Map` keeps the normalizer decoupled from the full `ChatCompletionResponse`/`ChatResponseMessage` shape (which carries many provider-specific fields and an `extra` flatten map), so it can't silently drop unknown passthrough fields on a re-serialize round-trip — a real concern on this codebase. The `String` allocations are bounded by the single non-stream response and only happen when a leading `<think>` block is actually present. The current approach was explicitly reviewed and approved; I'd rather not re-architect against that sign-off for a micro-optimization.

**Review by PierreLeGuen — APPROVED** [review]
Normalization of Chutes qwen `<think>` blocks into `reasoning_content` is correct and conservatively scoped — `split_leading_think_block` requires a start-anchored `<think>` with a matching close and returns `None` otherwise, so malformed input never loses content, and the guard treats `null`/empty `reasoning_content` as unset so minimax/GLM responses are unaffected. The normalizer runs only on the non-stream path after `transform_response_json`, the typed response is parsed from the same normalized bytes as `raw_bytes` so they can't diverge, and Chutes has no chat signatures, so no attestation invariant is touched. The earlier empty-`<think>`-block edge case is fixed in the head commit with a regression test.

No blocking findings.

Checks: `cargo test -p inference_providers --lib attested::chutes` (80 passed, 1 ignored live probe), `cargo clippy -p inference_providers --all-targets -- -D warnings` clean, `cargo fmt --check` clean; CI Lint and security_audit green, Test Suite still in progress at review time. Full e2e suite skipped (needs PostgreSQL; change is confined to the Chutes provider module).
